From 347a55643e690da9041cd58fe9289c794eac70a7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Sat, 21 Mar 2015 17:40:20 -0400
Subject: [PATCH] timedated: flip internal status after executing operation

timedated would set the internal status before calling out to systemd to do
the actual change. When the operation was refused because of a SELinux denial,
the state kept in timedated would get out of sync, and the second call from
timedatectl would appear to succeed.

https://bugzilla.redhat.com/show_bug.cgi?id=1014315

(cherry picked from commit 192b98b8fe73c8fb4bb3d6540deb93f5fb6eb9d2)
(cherry picked from commit fb14f86a7188f289dfc4081a6d83a5c9c7ce5a81)

[Amended to reverse the condition.]
---
 src/timedate/timedated.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/src/timedate/timedated.c b/src/timedate/timedated.c
index 08b604dcca..a568282141 100644
--- a/src/timedate/timedated.c
+++ b/src/timedate/timedated.c
@@ -286,7 +286,7 @@ static int context_start_ntp(Context *c, sd_bus *bus, sd_bus_error *error) {
         l = get_ntp_services();
         STRV_FOREACH(i, l) {
 
-                if (c->use_ntp)
+                if (!c->use_ntp)
                         r = sd_bus_call_method(
                                         bus,
                                         "org.freedesktop.systemd1",
@@ -337,7 +337,7 @@ static int context_enable_ntp(Context*c, sd_bus *bus, sd_bus_error *error) {
 
         l = get_ntp_services();
         STRV_FOREACH(i, l) {
-                if (c->use_ntp)
+                if (!c->use_ntp)
                         r = sd_bus_call_method(
                                         bus,
                                         "org.freedesktop.systemd1",
@@ -673,8 +673,6 @@ static int method_set_ntp(sd_bus *bus, sd_bus_message *m, void *userdata, sd_bus
         if (r == 0)
                 return 1;
 
-        c->use_ntp = ntp;
-
         r = context_enable_ntp(c, bus, error);
         if (r < 0)
                 return r;
@@ -683,6 +681,8 @@ static int method_set_ntp(sd_bus *bus, sd_bus_message *m, void *userdata, sd_bus
         if (r < 0)
                 return r;
 
+        c->use_ntp = ntp;
+
         log_info("Set NTP to %s", c->use_ntp ? "enabled" : "disabled");
 
         sd_bus_emit_properties_changed(bus, "/org/freedesktop/timedate1", "org.freedesktop.timedate1", "NTP", NULL);
